common/peerprovider/ringpopprovider/provider.go

```diff
@@ -0,0 +1,208 @@
+		status     int32
+		service    string
+		ringpop    *ringpop.Ringpop
+		bootParams *swim.BootstrapOptions
+		logger     log.Logger
+
+		smu         sync.RWMutex
+		subscribers map[string]chan<- *membership.ChangedEvent
+	}
+)
+
+const (
+	// roleKey label is set by every single service as soon as it bootstraps its
+	// ringpop instance. The data for this key is the service name
+	roleKey = "serviceName"
+)
+
+var _ membership.PeerProvider = (*Provider)(nil)
+
+func New(
+	service string,
+	config *Config,
+	channel tchannel.Channel,
+	logger log.Logger,
+) (*Provider, error) {
+	if err := config.validate(); err != nil {
+		return nil, err
+	}
+
+	rp, err := ringpop.New(config.Name, ringpop.Channel(channel.(*tcg.Channel)))
+	if err != nil {
+		return nil, fmt.Errorf("ringpop creation failed: %v", err)
+	}
+
+	discoveryProvider, err := newDiscoveryProvider(config, logger)
+	if err != nil {
+		return nil, fmt.Errorf("ringpop discovery provider: %v", err)
+	}
+
+	bootstrapOpts := &swim.BootstrapOptions{
+		MaxJoinDuration:  config.MaxJoinDuration,
+		DiscoverProvider: discoveryProvider,
+	}
+
+	return NewRingpopProvider(service, rp, bootstrapOpts, logger), nil
+}
+
+// NewRingpopProvider sets up ringpop based peer provider
+func NewRingpopProvider(service string,
+	rp *ringpop.Ringpop,
+	bootstrapOpts *swim.BootstrapOptions,
+	logger log.Logger,
+) *Provider {
+	return &Provider{
+		service:     service,
+		status:      common.DaemonStatusInitialized,
+		ringpop:     rp,
+		bootParams:  bootstrapOpts,
+		logger:      logger,
+		subscribers: map[string]chan<- *membership.ChangedEvent{},
+	}
+}
+
+// Start starts ringpop
+func (r *Provider) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&r.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	_, err := r.ringpop.Bootstrap(r.bootParams)
+	if err != nil {
+		r.logger.Fatal("unable to bootstrap ringpop", tag.Error(err))
+	}
+
+	// Get updates from ringpop ring
+	r.ringpop.AddListener(r)
+
+	labels, err := r.ringpop.Labels()
+	if err != nil {
+		r.logger.Fatal("unable to get ring pop labels", tag.Error(err))
+	}
+
+	if err = labels.Set(roleKey, r.service); err != nil {
+		r.logger.Fatal("unable to set ring pop labels", tag.Error(err))
+	}
+}
+
+// HandleEvent handles updates from ringpop
+func (r *Provider) HandleEvent(
+	event events.Event,
+) {
+	// We only care about RingChangedEvent
+	e, ok := event.(events.RingChangedEvent)
+	if !ok {
+		return
+	}
+
+	r.logger.Info("Received a ringpop ring changed event")
+	// Marshall the event object into the required type
+	change := &membership.ChangedEvent{
+		HostsAdded:   e.ServersAdded,
+		HostsUpdated: e.ServersUpdated,
+		HostsRemoved: e.ServersRemoved,
+	}
+
+	// Notify subscribers
+	r.smu.RLock()
+	defer r.smu.RUnlock()
+
+	for name, ch := range r.subscribers {
+		select {
+		case ch <- change:
+		default:
+			r.logger.Error("Failed to send listener notification, channel full", tag.Subscriber(name))
```

Comment:
is it really an error?  or is it expected to be so rare that it may be a sign of problems?